r-package/grf/R/boosted_regression_forest.R

```diff
@@ -0,0 +1,248 @@
+#'                            smaller weight than others in training the forest, i.e., the contribution
+#'                            of a given cluster to the final forest scales with the minimum of
+#'                            the number of observations in the cluster and samples_per_cluster.
+#' @param tune.parameters If true, NULL parameters are tuned by cross-validation; if false
+#'                        NULL parameters are set to defaults.
+#' @param num.fit.trees The number of trees in each 'mini forest' used to fit the tuning model.
+#' @param num.fit.reps The number of forests used to fit the tuning model.
+#' @param num.optimize.reps The number of random parameter values considered when using the model
+#'                          to select the optimal parameters.
+#' @param num.steps The number of boosting iterations. If NULL, selected by cross-validation
+#' @param max.steps The maximum number of boosting iterations to try when num.steps is not specified
+#' @param num.trees.tune If num.steps is NULL, the number of trees used to test a new boosting step when tuning
+#'        num.steps
+#'
+#' @return A trained regression forest object.
+#'
+#' @examples \dontrun{
+#' # Train a boosted regression forest.
+#' n = 50; p = 10
+#' X = matrix(rnorm(n*p), n, p)
+#' Y = X[,1] * rnorm(n)
+#' r.b.forest = boosted_regression_forest(X, Y)
+#'
+#' # Predict using the forest.
+#' X.test = matrix(0, 101, p)
+#' X.test[,1] = seq(-2, 2, length.out = 101)
+#' r.pred = predict(r.b.forest, X.test)
+#'
+#' # Predict on out-of-bag training samples.
+#' r.pred = predict(r.b.forest)
+#'
+#' #Check how many boosting iterations were used
+#' print(length(r.b.forest$forests))
+#' }
+#'
+#' @export
+boosted_regression_forest <- function(X, Y,
+                              sample.fraction = 0.5,
+                              mtry = NULL,
+                              num.trees = 500,
+                              num.threads = NULL,
+                              min.node.size = NULL,
+                              honesty = TRUE,
+                              honesty.fraction = NULL,
+                              ci.group.size = 2,
+                              alpha = NULL,
+                              imbalance.penalty = NULL,
+                              seed = NULL,
+                              clusters = NULL,
+                              samples_per_cluster = NULL,
+                              tune.parameters = FALSE,
+                              num.fit.trees = 10,
+                              num.fit.reps = 100,
+                              num.optimize.reps = 1000,
+                              num.steps = NULL,
+                              max.steps = 6,
+                              num.trees.tune = 10) {
+
+    boosted.forest = NULL
+    boosted.forest[["forests"]] = list()
+    forest.Y <- regression_forest(X, Y, sample.fraction = sample.fraction,
```

Comment:
I think this could be simplified a good deal. More on this on my overall comment.